### PR TITLE
Remove hard-coding of RAPIDS version where possible

### PR DESCRIPTION
**Pull request by KyleFromNVIDIA**
* Read `VERSION` file from CMake
* Read `ucxx.__version__` from docs build
* Read `VERSION` file from shell scripts
* Use variable substitution in Doxyfile
* Update `pyproject.toml` to use dynamic version
* Update `ci/release/update-version.sh`

Issue: https://github.com/rapidsai/build-planning/issues/15

### CODE REVIEW

**Review by pentschev — COMMENTED** [review]
It's not entirely clear to me whether this was taken into consideration, but note that version of UCXX is different from that of RAPIDS, so that might cause some problems. Perhaps this was already thought of, but wanted to raise awareness anyway.

**Comment by KyleFromNVIDIA** [comment]
Yes, that was taken into consideration, and it's why I didn't use the `rapids-version` and `rapids-version-major-minor` commands (among other deviations from the other repos.)

File: cpp/doxygen/Doxyfile [hunk 1]
```diff
@@ -38,7 +38,7 @@ PROJECT_NAME           = libucxx
 # could be handy for archiving the generated documentation or if some version
 # control system is used.
 
-PROJECT_NUMBER         = 0.37.00
+PROJECT_NUMBER         = $(UCXX_VERSION)
```
Comment (jakirkham) [inline cpp/doxygen/Doxyfile:41]: Where does this get set?

**Comment by jakirkham** [comment]
Think we may need to add some symlinks to the top-level `VERSION` file into Python projects

File: cpp/doxygen/Doxyfile (hunk 1 above)
Comment (KyleFromNVIDIA) [reply cpp/doxygen/Doxyfile:41]: See `cpp/CMakeLists.txt` and `ci/build_docs.sh`.

File: dependencies.yaml [hunk 2]
```diff
@@ -144,6 +144,7 @@ dependencies:
         packages:
           - setuptools>=64.0.0
           - tomli
+          - scikit-build-core[pyproject]>=0.7.0
```
Comment (bdice) [inline dependencies.yaml:147]: I don't think that `distributed-ucxx` uses `scikit-build-core`. See the `pyproject.toml`:
```toml
[build-system]
build-backend = "setuptools.build_meta"
```

Even if it did, `package[extras]` entries can't be used in conda. See `build_python_ucxx` above, and you'll notice that only `output_types: [requirements, pyproject]` use the `package[extras]` form.

**Review by bdice — CHANGES_REQUESTED** [review]

File: dependencies.yaml (hunk 2 above)
Comment (bdice) [reply dependencies.yaml:147]: This package probably needs to do the same as `dask_cudf`:

https://github.com/rapidsai/cudf/blob/241825a7b4db1713f44d8c298e08364b1eea9a32/python/dask_cudf/pyproject.toml#L60-L61

Comment (KyleFromNVIDIA) [reply dependencies.yaml:147]: You're quite right. I didn't realize `setuptools` had a utility to just read the version from a file.

**Comment by bdice** [comment]
Something seems to have broken ucxx wheel builds. It's failing with a `patchelf` error.

```
Fixed-up wheel written to /__w/ucxx/ucxx/python/final_dist/ucxx_cu11-0.37.0a36-cp310-cp310-manylinux_2_28_aarch64.whl
Archive:  /__w/ucxx/ucxx/python/final_dist/ucxx_cu11-0.37.0a36-cp310-cp310-manylinux_2_28_aarch64.whl
  inflating: repair_dist/ucxx_cu11.libs/libcudart-be0b6e85.so.11.8.89  
  inflating: repair_dist/ucxx_cu11.libs/libucp-d240e863.so.0.0.0  
  inflating: repair_dist/ucxx_cu11.libs/libucs-8ef1b6eb.so.0.0.0  
  inflating: repair_dist/ucxx_cu11.libs/libucm-ed183686.so.0.0.0  
  inflating: repair_dist/ucxx_cu11.libs/libuct-d5cf5765.so.0.0.0  
/__w/ucxx/ucxx/repair_dist/ucxx_cu11.libs /__w/ucxx/ucxx
/__w/ucxx/ucxx
/__w/ucxx/ucxx/repair_dist/ucxx_cu11.libs/ucx /__w/ucxx/ucxx
patchelf: missing filename
```

https://github.com/rapidsai/ucxx/actions/runs/8254849219/job/22581770392?pr=201#step:9:2121

**Comment by KyleFromNVIDIA** [comment]
See https://github.com/rapidsai/ucxx/pull/203 for a possible fix.

**Comment by jakirkham** [comment]
Now that #203 has landed, updating this PR to include that fix

**Review by bdice — APPROVED** [review]

**Review by AyodeAwe — APPROVED** [review]

**Review by pentschev — APPROVED** [review]
LGTM, thanks @KyleFromNVIDIA .

**Pull request merged**